Add copy resource interface, implement cmis resource copy with additional properties

wangf1122 opened the pull request:
The issue is when user copy cmisStore. All customized properties are not copied. We need additional properties copied as required by our system. So I added such interface.

josegar74 commented on core/src/main/java/org/fao/geonet/api/records/attachments/StoreUtils.java:
```diff
@@ -68,12 +68,9 @@ public static void copyDataDir(ServiceContext context, int oldMetadataId, int ne
-            }
+            //Copy from approved copy to working copy
+            boolean oldApproved = true;
+            store.copyResources(context, oldUuid, newUuid, visibility, oldApproved, newApproved);
```
Probably I'm missing something as haven't check in detail, but not sure if this is going to work with the `FilesystemStore`? In previous code calls ` store.putResource` that managed to copy the resources:

https://github.com/geonetwork/core-geonetwork/blob/17aa5333aa16546ca004e96d8429eaf9db95f931/core/src/main/java/org/fao/geonet/api/records/attachments/FilesystemStore.java#L187-L196

But in this new version calls `store.copyResources` that seem empty in `FilesystemStore`.

https://github.com/geonetwork/core-geonetwork/pull/5776/files#diff-27baa8be68aa76d7a5f25db314795edb869f980119dc5d781e82aea9d0f5611eR187-R191

ianwallen replied:
I agree, I believe this logic will break all other file stores.

It looks like the old logic should have been moved to the new "public void copyResources" within AbstractStore so that all other file stores would use that default logic. And all the code changes with "// TODO Auto-generated method stub" should be removed so that they inherit the default AbstractStore implementation.

wangf1122 replied:
I will try to move the old code to general block to all other store types.

wangf1122 replied:
Move StoreUtils.copyDataDir() implementations to AbstractStore.copyResources as default implementation for all other store type other than CmisStore. I tested locally for CMIS and FilesystemStore

ianwallen commented on core/src/main/java/org/fao/geonet/api/records/attachments/CMISStore.java:
```diff
@@ -426,6 +436,50 @@ public MetadataResourceContainer getResourceContainerDescription(final ServiceCo
+                if (sourceDocument instanceof Document) {
+                    // Get cmis properties from the source document
+                    Map<String, Object> sourceProperties = getSecondaryProperties(sourceDocument);
+                    putResource(context, targetUuid, sourceDocument.getName(), sourceDocument.getContentStream().getStream(), null, metadataResourceVisibility, targetApproved, sourceProperties);
```
There are regular CMIS properties and there are secondary properties.
You are passing the sourceProperties to the put resource function which are the secondary properties but you don't seem to be also including the regular properties in the list as well.

i.e. if the cmis:description property were set then it would get lots in the copy.

I believe the Document.copy command does copy the properties.

So you could do something along these lines.

```
                Document sourceDocument = sourceEntry.getValue();
                final int metadataId = canEdit(context, targetUuid, targetApproved);
                String destKey = getKey(context, targetUuid, metadataId, metadataResourceVisibility, sourceDocument.getName());
                int lastFolderDelimiterDestKeyIndex = destKey.lastIndexOf(cmisConfiguration.getFolderDelimiter());
                String parentDestFolderKey = destKey.substring(0, lastFolderDelimiterDestKeyIndex);
                Folder parentDestFolder = cmisUtils.getFolderCache(parentDestFolderKey, true);
                Document destDocument = sourceDocument.copy(parentDestFolder);
                setCmisMetadataUUIDSecondary(destDocument, getSecondaryProperties(sourceDocument), targetUuid);
```

Or an alternative could be to modify the sourceProperties so that it include both primary and secondary properties.

wangf1122 replied:
I tried. The backend content server is not happy with overwriting all properties. I can only do these secondary properties without causing crashing.

ianwallen replied:
I did some test and you are correct - the copy complains that document already exists.
I guess you can leave it as is for now until we identify exactly what properties would need to be copied

wangf1122 commented on core/src/main/java/org/fao/geonet/api/records/attachments/CMISStore.java:
```diff
@@ -189,18 +196,21 @@ public MetadataResource putResource(final ServiceContext context, final String m
                 doc.updateProperties(secondaryProperties);
+                if (MapUtils.isNotEmpty(additionalProperties)) {
+                    doc.updateProperties(additionalProperties,true);
+                }
```
These two properties can be merged together by map.putAll(anotherMap). So the update property will happen after.

ianwallen replied:
You need to call setCmisMetadataUUIDSecondary after the putAll properties.  Otherwise the changes done from the setCmisMetadataUUIDSecondary will be overwritten by the putAll.

wangf1122 replied:
Its been moved before the setCmisMetadataUUIDSecondary

wangf1122 commented on core/src/main/java/org/fao/geonet/api/records/attachments/CMISStore.java:
```diff
@@ -426,6 +436,50 @@ public MetadataResourceContainer getResourceContainerDescription(final ServiceCo
+
+            for (Map.Entry<String, Document> sourceEntry : sourceDocumentMap.entrySet()) {
+                Document sourceDocument = sourceEntry.getValue();
+                if (sourceDocument instanceof Document) {
```
I don't want to put too much constraints to assume this is a document for sure. If it's not a document, then it will do just skip it and look for the next cmis object. This code will not affect any performance by just for more safety.

Anything can be possible happening at the backend content server or its script. We can do more thing maybe put a warning/error here to log it.

ianwallen replied:
I don't think you understand my point. 
The following identifies that sourceDocument is a "Document" object.
`Document sourceDocument`
So it is not possible for the following to fail the condition.
`if (sourceDocument instanceof Document) {`
I'm surprised that the compiler is not complaining about this by stating that the condition will always be true.

wangf1122 replied:
oh. I get it. Thanks

wangf1122 commented on core/src/main/java/org/fao/geonet/api/records/attachments/CMISStore.java:
```diff
@@ -426,6 +436,50 @@ public MetadataResourceContainer getResourceContainerDescription(final ServiceCo
+
+    private Map<String, Object> getSecondaryProperties(Document document) {
+        String secondaryPropertyId=null;
+        for (Property<?> property:document.getProperties()) {
```
Good catch

wangf1122 commented on core/src/main/java/org/fao/geonet/api/records/attachments/CMISStore.java:
```diff
@@ -426,6 +436,50 @@ public MetadataResourceContainer getResourceContainerDescription(final ServiceCo
+        Map<String, Object> properties = new HashMap<>();
+        if (!StringUtils.isEmpty(secondaryPropertyId)) {
+            for (Property<?> property:document.getProperties()) {
+                if (property.getId().contains(secondaryPropertyId) && property.getValue()!=null) {
```
yes. We don't need that ":" as long as it starts with secondaryPropertyId. It's fine

ianwallen replied:
"test:propertyname" and test2:property name both start with test but only one is part of the test aspect.  If the ":" is not part of the check then there is a possibility that it gets the wrong value.
But I do admit that the changes of something like this happening is very low.

josegar74 commented:
I have done this test, not sure if doesn't work due to an issue in the PR or there's some code missing in the 4.0.x related to the workflow.

1) Enable the workflow.

2) Create a metadata and add a thumbnail  and a file and approve / publish it.

3) Edit the metadata and in the draft copy remove the file (from the online resource and the store) and add a new one.

Until this point the data directory is fine for published and draft versions.

4) Approve the draft copy --> In the published metadata data directory there's only the file for the thumbnail, the new file added in previous step is missed (it is listed as online resource, but the link doesn't work as the file is missing).

ianwallen commented on core/src/main/java/org/fao/geonet/api/records/attachments/AbstractStore.java:
```diff
@@ -206,6 +206,16 @@ public MetadataResource patchResourceStatus(final ServiceContext context, final
 
+    @Override
+    public void copyResources(ServiceContext context, String sourceUuid, String targetUuid, MetadataResourceVisibility metadataResourceVisibility, boolean sourceApproved, boolean targetApproved) throws Exception {
+        final List<MetadataResource> resources = getResources(context, sourceUuid, metadataResourceVisibility, null, true);
```
I believe "true" should be using variable sourceApproved

wangf1122 replied:
yes. This causes @josegar74  test failure. That's why when he copies from working copy -> approved copy. The files always got wiped out because this hardcoded flag "true" always gets sources as approved copy. So the File.copy method copies the input stream of the same file to the same location and causing issues. Now it should get fixed.

ianwallen commented on core/src/main/java/org/fao/geonet/api/records/attachments/AbstractStore.java:
```diff
@@ -206,6 +206,16 @@ public MetadataResource patchResourceStatus(final ServiceContext context, final
+    public void copyResources(ServiceContext context, String sourceUuid, String targetUuid, MetadataResourceVisibility metadataResourceVisibility, boolean sourceApproved, boolean targetApproved) throws Exception {
+        final List<MetadataResource> resources = getResources(context, sourceUuid, metadataResourceVisibility, null, true);
+        for (MetadataResource resource: resources) {
+            try (Store.ResourceHolder holder = getResource(context, sourceUuid, metadataResourceVisibility, resource.getFilename(), true)) {
```
I believe "true" should be using variable sourceApproved

wangf1122 replied:
fixed

ianwallen commented on core/src/main/java/org/fao/geonet/api/records/attachments/StoreUtils.java:
```diff
@@ -133,20 +130,8 @@ public static void replaceDataDir(ServiceContext context, String sourceUuid, Str
             List<MetadataResource> targetAddResources = exceptMetadataResource(sourceResources, targetResources);
             List<MetadataResource> targetUpdateResources = exceptMetadataResource(targetResources, targetDeleteResources);
```
Remove these variables as they are no longer being used.

wangf1122 replied:
removed

ianwallen reviewed (commented):
@wangf1122 
I found a few other small bugs - but I don't think it is related to the issue @josegar74  found.

wangf1122 commented:
> I have done this test, not sure if doesn't work due to an issue in the PR or there's some code missing in the 4.0.x related to the workflow.
> 
> 1. Enable the workflow.
> 2. Create a metadata and add a thumbnail  and a file and approve / publish it.
> 3. Edit the metadata and in the draft copy remove the file (from the online resource and the store) and add a new one.
> 
> Until this point the data directory is fine for published and draft versions.
> 
> 1. Approve the draft copy --> In the published metadata data directory there's only the file for the thumbnail, the new file added in previous step is missed (it is listed as online resource, but the link doesn't work as the file is missing).

I couldn't reproduce it from my end. I can always get the file copied. Did you use file system resource directly?

Here is my test:
1) Resource name vignette_ladyhammond_s.png metadata internal ID = 3203
![image](https://user-images.githubusercontent.com/74916635/124511928-37c60a00-dda5-11eb-8b95-596ccc333ee2.png)

2) Approved it then made a working copy. Deleted vignette_ladyhammond_s.png and added earth.jpg.
![image](https://user-images.githubusercontent.com/74916635/124512529-950e8b00-dda6-11eb-89f6-e5597db544e4.png)

3) Approved the working copy. Check my backend cmis source. The folder of 3203 contains only the "earth.jpg"
![image](https://user-images.githubusercontent.com/74916635/124512622-d2731880-dda6-11eb-8d28-133197440676.png)

josegar74 commented on core/src/main/java/org/fao/geonet/api/records/attachments/StoreUtils.java:
```diff
@@ -130,23 +127,9 @@ public static void replaceDataDir(ServiceContext context, String sourceUuid, Str
-                }
-            }
+            // copy records from source to target
+            store.copyResources(context, sourceUuid, targetUuid, visibility, sourceApproved, targetApproved);
```
I'm not sure about this change. Seems the `Add new records` /  `update new records` blocks have been replaced by previous line, but I don't see the original logic is preserved.

ianwallen replied:
@josegar74 
I had concerns as well however after reviewing it, I believe it is OK.

The add and the update were basically doing the same thing by calling 
`store.putResource(context, targetUuid, holder.getPath(), visibility, targetApproved);`
for all the files that existed in the source folder.
The main difference in the previous logic is that it was separating them into insert and updates so that we could potentially call different methods in the future. But currently,  different logic currently does not exists for inserts and updates, so it looks like @wangf1122 simply did them all for one function.

The deletes are currently the only one that need to be done separately.

So the main thing that we are loosing from this change is the separation of the insert/updates - but since they were calling the same function, it was not really used.  We can always change it back in the future, if there is a need for using different functions for the inserts and updates.

josegar74 replied:
Indeed, thanks for the clarification.
